### ISSUE TITLE
Implement Comprehensive Image Data Augmentation Framework

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#578]
## Overview

Merge of #389 (Standard Image Data Augmentation) and #390 (Advanced Augmentation Mixup/CutMix) into a comprehensive, production-ready image augmentation framework.

### 1. New `AugmentationBase<T>` (Not extending TransformerBase)

Augmentations are fundamentally different from preprocessing transforms:
- **Stochastic**: Same input produces different outputs each call
- **Training-only**: Often disabled during inference
- **Probabilistic**: Each augmentation has application probability
- **Composable**: Support random/sequential ordering

### 2. New `ImageTensor<T>` Data Type

Specialized wrapper with image-specific metadata:
- Shape conventions: `[N, C, H, W]` (batch, channels, height, width)
- Color space tracking (RGB, BGR, Grayscale, etc.)
- Channel order metadata
- Normalization state tracking

### 3. Callback/Event System for Soft Labels (Mixup/CutMix)

Integration with existing event infrastructure for label mixing:
- Augmentation emits events when labels need modification
- Training loop subscribes to handle soft labels
- Supports custom mixing strategies

### 4. Configurable Pipeline Ordering

`AugmentationPipeline` supports multiple ordering strategies:
- `Sequential`: Fixed order as defined
- `Random`: Shuffle order each batch (RandAugment-style)
- `Custom`: User-defined ordering logic

---

#### 1.1 Core Interfaces and Base Classes
**Location**: `src/Images/Augmentation/`

```
src/Images/Augmentation/
├── Interfaces/
│   ├── IImageAugmentation.cs        # Core augmentation contract
│   ├── IAugmentationPipeline.cs     # Pipeline composition
│   └── IAugmentationEvent.cs        # Event system for soft labels
├── Base/
│   ├── AugmentationBase.cs          # Abstract base with common logic
│   ├── GeometricAugmentationBase.cs # Base for spatial transforms
│   └── ColorAugmentationBase.cs     # Base for color transforms
├── ImageTensor.cs                    # Image-specific tensor wrapper
├── AugmentationPipeline.cs           # Composable pipeline
└── AugmentationOptions.cs            # Configuration options
```

**Key Interface:**
```csharp
public interface IImageAugmentation<T>
{
    ImageTensor<T> Apply(ImageTensor<T> image, Random? random = null);
    bool IsTrainingOnly { get; }
    double Probability { get; }
    string Name { get; }
    
    // Event for augmentations that modify labels (Mixup, CutMix)
    event EventHandler<LabelMixingEventArgs<T>>? OnLabelMixing;
}
```

### Phase 2: Basic Geometric Augmentations (CRITICAL)

**Location**: `src/Images/Augmentation/Geometric/`

| Augmentation | Priority | Description |
|--------------|----------|-------------|
| `RandomCrop` | CRITICAL | Crop random region, with padding options |
| `RandomHorizontalFlip` | CRITICAL | Flip with configurable probability |
| `RandomVerticalFlip` | HIGH | Vertical flip |
| `RandomRotation` | HIGH | Rotate by random angle within range |
| `RandomResizedCrop` | HIGH | Crop + resize (scale augmentation) |
| `RandomAffine` | MEDIUM | Combined rotation, translation, scale, shear |
| `CenterCrop` | MEDIUM | Deterministic center crop |
| `Resize` | MEDIUM | Deterministic resize |
| `Pad` | MEDIUM | Padding with various modes (reflect, constant, etc.) |

### Phase 3: Color/Intensity Augmentations (HIGH)

**Location**: `src/Images/Augmentation/Color/`

| Augmentation | Priority | Description |
|--------------|----------|-------------|
| `Normalize` | CRITICAL | Channel-wise mean/std normalization |
| `ColorJitter` | HIGH | Random brightness, contrast, saturation, hue |
| `GaussianBlur` | MEDIUM | Random Gaussian blur with kernel size range |
| `GaussianNoise` | MEDIUM | Add Gaussian noise |
| `RandomGrayscale` | MEDIUM | Convert to grayscale with probability |
| `RandomInvert` | LOW | Invert colors with probability |
| `RandomSolarize` | LOW | Solarize above threshold |
| `RandomPosterize` | LOW | Reduce bits per channel |

### Phase 4: Modern Regularization (HIGH)

**Location**: `src/Images/Augmentation/Regularization/`

| Augmentation | Priority | Description |
|--------------|----------|-------------|
| `Mixup` | HIGH | Linear interpolation of image pairs + soft labels |
| `CutMix` | HIGH | Cut-and-paste regions + proportional labels |
| `Cutout` | MEDIUM | Random erasing of rectangular regions |
| `RandomErasing` | MEDIUM | Erase with random values (not just zero) |
| `GridMask` | LOW | Grid-based masking pattern |

**Soft Label Event System:**
```csharp
public class LabelMixingEventArgs<T> : EventArgs
{
    public Vector<T> OriginalLabels { get; }
    public Vector<T> MixedLabels { get; set; }
    public T MixingLambda { get; }
    public int[] MixedIndices { get; }  // For CutMix: which samples were mixed
}
```

### Phase 5: Auto-Augmentation Policies (MEDIUM)

**Location**: `src/Images/Augmentation/Policies/`

| Policy | Priority | Description |
|--------|----------|-------------|
| `RandAugment` | HIGH | Random selection from augmentation pool with magnitude |
| `TrivialAugment` | MEDIUM | Simplified RandAugment (single augmentation per image) |
| `AutoAugment` | MEDIUM | Learned policies (ImageNet, CIFAR, SVHN presets) |
| `AugMax` | LOW | Adversarial augmentation selection |

---

### Deterministic Seeding
- All augmentations accept optional `Random` or seed
- Use `RandomHelper.CreateSeededRandom()` / `RandomHelper.CreateSecureRandom()`
- Pipeline can set global seed for reproducibility

### GPU Acceleration (Future)
- Design interface to support GPU backends
- Initial implementation CPU-only with clear extension points
- Consider `IComputeBackend` abstraction

### Performance
- Minimize allocations (reuse buffers where possible)
- Support in-place operations where safe
- Batch-level operations for Mixup/CutMix

### Parity Goals
- Match torchvision.transforms API semantics
- Match albumentations performance characteristics
- Exceed both with .NET-native optimizations

---

## Success Criteria

- [ ] `ImageTensor<T>` with full metadata support
- [ ] `AugmentationBase<T>` with probability and training mode
- [ ] `AugmentationPipeline` with sequential/random ordering
- [ ] All 9 geometric augmentations implemented
- [ ] All 8 color augmentations implemented
- [ ] Mixup and CutMix with event-based soft labels
- [ ] RandAugment and TrivialAugment policies
- [ ] Comprehensive unit tests with visual verification
- [ ] Deterministic seeding verified
- [ ] Integration example with neural network training

---

## References

- torchvision.transforms: https://pytorch.org/vision/stable/transforms.html
- albumentations: https://albumentations.ai/docs/
- RandAugment paper: https://arxiv.org/abs/1909.13719
- Mixup paper: https://arxiv.org/abs/1710.09412
- CutMix paper: https://arxiv.org/abs/1905.04899

---

Supersedes: #389, #390
